File: tools/fuckGFW.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
Update hosts for *nix
Author: cloud@txthinking.com
Version: 0.0.1
Date: 2012-10-24 14:35:39
'''

import urllib2
import os
import sys

HOSTS_PATH = "/etc/hosts"
HOSTS_SOURCE = "http://tx.txthinking.com/hosts"
SEARCH_STRING = "#TX-HOSTS"

def GetRemoteHosts(url):
    f = urllib2.urlopen(url, timeout=5)
    hosts = [line for line in f]
    f.close()
    return hosts

def main():
    try:
        hosts = GetRemoteHosts(HOSTS_SOURCE)
    except IOError:
        print "Could't connect to %s. Try again." % HOSTS_SOURCE
        sys.exit(1)

    yours = ""
    if os.path.isfile(HOSTS_PATH):
        f = open(HOSTS_PATH, "r")
        for line in f:
            if SEARCH_STRING in line:
                break
            yours += line
        f.close()
        os.rename(HOSTS_PATH, HOSTS_PATH + ".BAK")
    yours += SEARCH_STRING + "\n"

    fp = open(HOSTS_PATH, "w")
    fp.write(yours)
    fp.writelines(hosts)
    fp.close()

    print "Success"

if __name__ == "__main__":
    main()

File: tools/fuckGFW.go
//
// Update hosts for all OS
//

package main

import (
    "os"
    "io"
    "bufio"
    "net/http"
    "time"
    "bytes"
    "io/ioutil"
    "runtime"
    "fmt"
)

const (
    SEARCH_STRING string = "#TX-HOSTS"
    HOSTS_SOURCE string = "http://freedom.txthinking.com/hosts"
)

func Exit(message string){
    fmt.Println(message)
    time.Sleep(3 * time.Second)
    os.Exit(0)
}

func main(){
    BR := "\n"
    HOSTS_PATH := "/etc/hosts"
    if runtime.GOOS == "windows"{
        BR = "\r\n"
        HOSTS_PATH = os.Getenv("SYSTEMROOT")+"\\system32\\drivers\\etc\\hosts"
    }

    r, err := http.Get(HOSTS_SOURCE)
    if err != nil {
        Exit(err.Error() + "Try again.")
    }
    data, err := ioutil.ReadAll(r.Body)
    if err != nil {
        Exit(err.Error())
    }
    hosts := string(bytes.Replace(data, []byte("\n"), []byte(BR), -1))

    yours := ""
    f, err := os.OpenFile(HOSTS_PATH, os.O_RDONLY, 0444)
    if err == nil {
        bnr := bufio.NewReader(f)
        for{
            line, _, err := bnr.ReadLine()
            if bytes.Equal(line,[]byte(SEARCH_STRING)) || err == io.EOF{
                break
            }
            yours += string(line) + BR
        }
        err = f.Close()
        if err != nil {
            Exit(err.Error())
        }
        _ = os.Remove(HOSTS_PATH+".BAK")
        _ = os.Rename(HOSTS_PATH, HOSTS_PATH+".BAK")
    }
    yours += SEARCH_STRING + BR

    f, err = os.OpenFile(HOSTS_PATH, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, 0644)
    if err != nil {
        Exit(err.Error())
    }
    _, err = f.WriteString(yours + hosts)
    if err != nil {
        Exit(err.Error())
    }
    err = f.Close()
    if err != nil {
        Exit(err.Error())
    }

    Exit("Success!")
}
